server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1172,10 +1172,22 @@ private void finalizeSnapshotEntry(SnapshotsInProgress.Entry entry, Metadata met
             if (entry.isClone()) {
                 threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(
                         ActionRunnable.supply(metadataListener, () -> {
-                            final Metadata.Builder metaBuilder = Metadata.builder(repo.getSnapshotGlobalMetadata(entry.source()));
+                            final Metadata existing = repo.getSnapshotGlobalMetadata(entry.source());
+                            final Metadata.Builder metaBuilder = Metadata.builder(existing);
+                            final Set<Index> existingIndices = new HashSet<>();
                             for (IndexId index : entry.indices()) {
-                                metaBuilder.put(repo.getSnapshotIndexMetaData(repositoryData, entry.source(), index), false);
+                                final IndexMetadata indexMetadata = repo.getSnapshotIndexMetaData(repositoryData, entry.source(), index);
+                                existingIndices.add(indexMetadata.getIndex());
+                                metaBuilder.put(indexMetadata, false);
                             }
+                            // remove those data streams from metadata for which we are missing indices
+                            Map<String, DataStream> dataStreamsToCopy = new HashMap<>();
+                            for (Map.Entry<String, DataStream> dataStreamEntry : existing.dataStreams().entrySet()) {
```

Comment:
An alternative here would be to just drop all data streams from the cloned snapshot (since we are cloning by index and not by data-stream). I figured this was potentially nicer for users that might want to use the functionality to clone all indices except for a subset of them to e.g. selectively removed data from existing snapshots. Since we added copying the global metadata to cloning for the same reason I did it this way.